src/BlockHeader.js

```diff
@@ -0,0 +1,177 @@
+import {connect} from 'react-redux';
+import {
+  getLabel,
+  getNestedBlockDefinition, isNA, isSimpleLayout,
+  isStruct, structValueToObject, triggerCustomEvent
+} from './processing/utils';
+import {toggleBlock} from './actions';
+import BlockActions from './BlockActions';
+import {refType} from './types';
+
+
+@connect((state, props) => {
+  const {fieldId, blockId} = props;
+  const fieldData = state[fieldId];
+  const blocks = fieldData.blocks;
+  const block = blocks[blockId];
+  const blockDefinition = getNestedBlockDefinition(state, fieldId, blockId);
+  const value = block.value;
+  return {
+    isSimpleLayout: isSimpleLayout(blockDefinition, fieldData.isMobile),
+    blockDefinition,
+    icons: fieldData.icons,
+    value: isStruct(blockDefinition) ?
+             structValueToObject(state, fieldId, value)
+             :
+             value,
+  };
+}, (dispatch, props) => {
+  const {fieldId, blockId} = props;
+  return bindActionCreators({
+    toggleBlock: () => toggleBlock(fieldId, blockId),
+  }, dispatch);
+})
+class BlockHeader extends React.Component {
+  static propTypes = {
+    fieldId: PropTypes.string.isRequired,
+    blockId: PropTypes.string.isRequired,
+    collapsibleBlock: PropTypes.bool,
+    sortableBlock: PropTypes.bool,
+    canDuplicate: PropTypes.bool,
+    dragHandleRef: refType,
+    dragHandleProps: PropTypes.object,
+  };
+
+  static defaultProps = {
+    collapsibleBlock: true,
+    sortableBlock: true,
+    canDuplicate: true,
+  };
+
+  get icon() {
+    const {blockDefinition} = this.props;
+    if (blockDefinition.icon !== undefined) {
+      return <span className="type-icon" dangerouslySetInnerHTML={
+        {__html: blockDefinition.icon}
+      } />;
+    }
+  }
+
+  get title() {
+    const {title, blockDefinition, value} = this.props;
+    if (isStruct(blockDefinition)) {
+      if ((title !== undefined) && (title !== null)) {
+        return title;
+      }
+      if (blockDefinition.titleTemplate !== undefined) {
+        let hasVariables = false;
+        let isEmpty = true;
+        let renderedTitle = blockDefinition.titleTemplate.replace(
+          /\${([^}]+)}/g, (match, varName) => {
+            let childValue = value[varName];
+            if (isNA(childValue)) {
+              childValue = '';
+            } else if (childValue !== '') {
+              isEmpty = false;
+            }
+            hasVariables = true;
+            return childValue;
+          });
+        if (!hasVariables || !isEmpty) {
+          return renderedTitle;
+        }
+      }
+    }
+    return null;
+  }
+
+  get titleAndType() {
+    const title = this.title;
+    const icon = this.icon;
+    const blockType = (
+      <span className="block-type">
+        {getLabel(this.props.blockDefinition)}
+      </span>
+    );
+    if (title === null) {
+      if (icon === null) {
+        return blockType;
+      }
+
+      return (
+        <span>
+          <h3>{icon}</h3>
```

Comment:
It doesn't seem like h3 would be the appropriate element for an icon. Also feels strange to have an `h3` in a `span`.

Reply:
Also icon should be wrapped in a `aria-hidden` element so screen-readers ignore it.